libclamav/ole2_extract.c

```diff
@@ -586,6 +586,281 @@ static cl_error_t handler_enum(ole2_header_t *hdr, property_t *prop, const char
 static cl_error_t handler_otf_encrypted(ole2_header_t *hdr, property_t *prop, const char *dir, cli_ctx *ctx, void *handler_ctx);
 static cl_error_t handler_otf(ole2_header_t *hdr, property_t *prop, const char *dir, cli_ctx *ctx, void *handler_ctx);
 
+/*
+ * Compare strings ignoring case.
+ * This is a somewhat special case, since name is actually a utf-16 encoded string, stored
+ * in a char * with a known size of 64 bytes, so we can avoid a 'alloc since the size is
+ * so small.  See https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-cfb/60fe8611-66c3-496b-b70d-a504c94c9ace
+ *
+ * @param name:                     'name' from property_t struct
+ * @param name_size:                'name_size' from property_t struct
+ * @param keyword:                  Known value we are looking for
+ *
+ * @return int:                     Return '0' if the values are equivalent, something else otherwise.
+ */
+static int ole2_cmp_name(const char *const name, uint32_t name_size, const char *const keyword)
+{
+    char decoded[64];
+    uint32_t i = 0, j = 0;
+
+    if (64 < name_size || name_size % 2) {
+        return -1;
+    }
+
+    memset(decoded, 0, sizeof(decoded));
+    for (i = 0, j = 0; i < name_size; i += 2, j++) {
+        decoded[j] = ((unsigned char)name[i + 1]) << 4;
+        decoded[j] += name[i];
+    }
+
+    return strcasecmp(decoded, keyword);
+}
+
+/*
+ * File Information Block Base.
+ * Naming is consistent with
+ * https://learn.microsoft.com/en-us/openspecs/office_file_formats/ms-doc/26fb6c06-4e5c-4778-ab4e-edbf26a545bb
+ * */
+typedef struct fib_base_type {
+    uint16_t wIdent __attribute__((packed));
+    uint16_t nFib;
+    uint16_t unused;
+    uint16_t lid;
+    uint16_t pnNext;
+    uint16_t ABCDEFGHIJKLM;
+    uint16_t nFibBack;
+    uint32_t lKey;
+    uint8_t envr;
+    uint8_t NOPQRS;
+    uint16_t reserved3;
+    uint16_t reserved4;
+    uint32_t reserved5;
+    uint32_t reserved6;
+} fib_base_t;
```

Comment:
I think this struct should be packed with the three different ways.

Reply:
agree, done